### PR TITLE
Test the build options

### PR DESCRIPTION
**Pull request by bzsolt**
Now the precommit checks only the {release,debug} native targets and unittests
with all combinations. Because testing on every target needs a lots of efforts,
added a new target called "precommit_heavy".
This target checks all targets with every combinations.

JerryScript-DCO-1.0-Signed-off-by: Zsolt Borbély zsborbely.u-szeged@partner.samsung.com

### CODE REVIEW

**Comment by bzsolt** [comment]
The precommit fail related to #762.

**Comment by LaszloLango** [comment]
@bzsolt, please rebase to the current master

File: CMakeLists.txt [hunk 1]
```diff
@@ -113,9 +113,9 @@ project (Jerry C ASM)
   set(MCU_SCRIPT_GENERATED_HEADER ${CMAKE_BINARY_DIR}/generated.h)
 
 # Should we use external libc?
- if(DEFINED USE_COMPILER_DEFAULT_LIBC AND USE_COMPILER_DEFAULT_LIBC STREQUAL "YES")
+ if(DEFINED COMPILER_DEFAULT_LIBC AND COMPILER_DEFAULT_LIBC STREQUAL "ON")
```
Comment (LaszloLango) [inline CMakeLists.txt]: Why did you rename this? Are you sure you changed all reference of `USE_COMPILER_DEFAULT_LIBC`?

**Comment by zherczeg** [comment]
LGTM

**Comment by zherczeg** [comment]
Landed in b8b587f76c41fcd3b46550642a64d0fad34f5610.

**Comment by zherczeg** [comment]
Sorry wrong PR :)

File: Makefile [hunk 2]
```diff
@@ -332,6 +349,8 @@ precommit: prerequisites
 	$(Q)+$(MAKE) --no-print-directory clean
 	$(Q) echo "Running checks..."
 	$(Q)+$(MAKE) --no-print-directory check-signed-off check-vera check-cppcheck
+	$(Q) echo "...testing build-options..."
+	$(Q)+$(MAKE) --no-print-directory test-buildoptions
```
Comment (akiss77) [inline Makefile]: I'm concerned about performance here. I quite often just run `make` (== `make precommit`) to see that everything is all right (checks, builds, tests). However, with this addition, this will perform all the builds with various options too. And this will call out to a shell script, which sequentially walks through targets and options, and then calls back to make again, thus loosing all parallelization potential. (Not to mention that ninja builds are way faster than non-ninja builds, but even if I run `make precommit NINJA=1`, because of the make->shell->make calls, this setting will be lost and loose ninja as well.) Is there no way to get everything done within the Makefile?

File: CMakeLists.txt (hunk 1 above)
Comment (bzsolt) [inline CMakeLists.txt]: Yes, I've changed all occurences of ```USE_COMPILER_DEFAULT_LIBC```.
The ```YES/NO``` -> ```ON/OFF``` modification just simplify the testing-mechanism, because only this option use ```YES/NO```, the others use ```ON/OFF```. 
In my opinion, the ```ON/OFF``` is more obvious without the ```USE_``` prefix.

Comment (LaszloLango) [inline CMakeLists.txt]: OK, thanks for explanation.

File: Makefile (hunk 2 above)
Comment (bzsolt) [inline Makefile]: You're right, I will check the way to move this logic to the Makefile.

**Comment by bzsolt** [comment]
I've updated the PR, @akiss77 could you check it please?

File: Makefile [hunk 3]
```diff
@@ -311,6 +324,19 @@ check-vera:
 check-cppcheck:
 	$(Q) $(call SHLOG,./tools/check-cppcheck.sh,$(OUT_DIR)/cppcheck.log,Cppcheck)
 
+.PHONY: test-buildoptions
+test-buildoptions:
+	$(Q) for target in $(JERRY_BUILD_OPTIONS_TEST_TARGETS); do \
+           if [[ $$target =~ "mcu" ]]; then \
+             current_options=$$BUILD_OPTIONS_TEST_MCU; \
```
Comment (akiss77) [inline Makefile]: We are in the Makefile, so we can use `$(BUILD_OPTIONS_TEST_MCU)`, we don't have to rely on the make variable being exported to a shell variable.

File: Makefile [hunk 4]
```diff
@@ -311,6 +324,19 @@ check-vera:
...
+           else \
+             current_options=$$BUILD_OPTIONS_TEST_NATIVE; \
```
Comment (akiss77) [inline Makefile]: ditto

File: Makefile (hunk 2 above)
Comment (akiss77) [inline Makefile]: It should be discussed and decided whether we want this to be included in the precommit target or have a separate target for the "heavy" checks (as was the case in a previous version, if I recall correctly). What's the impact of this addition on the precommit time? (Ofc, it can also be a valid answer that we don't care and want this tested, whatever it takes.)

File: Makefile [hunk 5]
```diff
@@ -311,6 +324,19 @@ check-vera:
...
+           fi; \
+           for option in $$current_options; do \
+             $(MAKE) $$option=ON $$target || exit 1; \
+           done \
+         done
```
Comment (akiss77) [inline Makefile]: What still bothers me is that these for loops are still sequential. Cannot we declare a make target for each of the `$(MAKE) $$option=ON $$target` commands and have all those targets be a dependecy of test-buildoptions? Then make would take care of parallelization, if it's available on the system. (Ex: test-js.*.* targets)

File: Makefile (hunk 2 above)
Comment (bzsolt) [inline Makefile]: With this version the runtime is more than twice before.
But now when the option-testing running in parallel and the build-option step included into the precommit process, the runtime only increased to ~5:55 minutes from ~3:25 on my local machine. So basically option-testing takes ~2:30 minutes.
What do you think, is this slow-down acceptable or not?

File: Makefile [hunk 6]
```diff
@@ -196,7 +216,9 @@ define WRITE_TOOLCHAIN_CONFIG
 endef
 
 .PHONY: $(BUILD_DIR)/$(NATIVE_SYSTEM)/toolchain.config
-$(BUILD_DIR)/$(NATIVE_SYSTEM)/toolchain.config:
+.PHONY: $(BUILD_DIR)/$(NATIVE_SYSTEM)/unittests/toolchain.config
```
Comment (bzsolt) [inline Makefile]: The cause of this (and related) modifications is that the parallel native builds can interfere with the unittests build.

**Comment by akiss77** [comment]
LGTM (please rebase)

**Comment by LaszloLango** [comment]
LGTM
@zherczeg, please check the changes.

**Comment by zherczeg** [comment]
LGTM
